README.md
# fmp4streamer
[What is fmp4streamer](#what-is-fmp4streamer) |
[How does it work](#how-does-it-work) |
[Capabilities](#capabilities) |
[Installation](#installation) |
[Running](#running) |
[Viewing](#viewing) |
[Configuration](#configuration) |
[Roadmap](#roadmap) |
[Motivation](#motivation) |
[Changelog](https://github.com/soyersoyer/fmp4streamer/blob/main/CHANGELOG.md)

# What is fmp4streamer
The fmp4streamer is a Python application designed to stream hardware encoded H.264 from a V4L2 Linux video device directly to a browser.

# How does it work
This python script setups the V4L2 device, reads the h264 stream from it, adds some fMP4 (fragmented MP4) header and serves it via HTTP or HTTPS. It works with only one html5 video tag, no js needed. It's pretty lightweight.

# Capabilities
- Stream to multiple clients simultaneously (usually only limited by your network connection) 
- Support any resolution and framerate the video device can capture
- Works in any Raspberry Pi with a camera module
- Able to handle high framerate (60-90 fps) streams
- Able to stream to iPhone and Safari via HLS.
- Very low cpu utilization, even when streaming using SSL

# Installation
   ```
   wget https://github.com/soyersoyer/fmp4streamer/archive/refs/tags/v3.0.3.zip
   unzip v3.0.3.zip
   mv fmp4streamer-3.0.3 fmp4streamer
   ```
# Installation as a system service
   ```
   git clone http://www.github.com/mclendon99/fmp4streamer.git
   ```

- Set up the config file
   ```
   cd fmp4streamer
   cp fmp4streamer.conf.dist fmp4streamer.conf
   ```
- Edit the fmp4streamer.conf file to point to your certificate/key if using HTTPS. (See below).
- Make any other changes as needed, e.g. port number or frame-rate...

- Install the config file
   ```
   sudo mkdir /etc/fmp4streamer
   sudo cp fmp4streamer.conf /etc/fmp4streamer
   sudo cp fmp4streamer.py /etc/fmp4streamer
   sudo cp bmff.py /etc/fmp4streamer
   sudo cp v4l2camera.py /etc/fmp4streamer
   sudo cp v4l2.py /etc/fmp4streamer
   ```
- Install the system service
   ```
   sudo cp fmp4streamer.system-service /lib/systemd/system/fmp4streamer.service
   sudo chmod 0644 /lib/systemd/system/fmp4streamer.service
   ```

   To use https, one must obtain a certificate from a registered domain.
   GoDaddy has cheap domains and LetsEncrypt will issue a certificate
   manually after installation of snap and letsencrypt, e.g. the following will manually issue a certificate:
   ```
   certbot certonly --authenticator dns-godaddy --dns-godaddy-propogation-seconds 600 --keep-until-expiring  --expand --server https://acme-v02.api.letencrypt.org/directory -d '.your-registered.domain' -d '*.your-registered.domain'
   ```
   and then put the challenge in the domain record per the instructions 
   and your keys will appear in /etc/letsencrypt/live/your-registered.domain/

   Point to the fullchain.pem certificate and the privkey.pem in the fmp4streamer.conf file.

# Running 
- from the terminal
    ```
    cd fmp4streamer
    python3 fmp4streamer.py
    ```
- At startup. Now installed as a system service.
    ```
    cd fmp4streamer
    mkdir -p ~/.config/systemd/user
    cp fmp4streamer.service ~/.config/systemd/user/
    systemctl --user enable fmp4streamer
    systemctl --user start fmp4streamer
    loginctl enable-linger pi
    ```

    watch the logs
    ```
    systemctl --user status fmp4streamer
    journalctl --user-unit fmp4streamer
    ```

# Running as a system service (one time only)
    ```
    sudo systemctl daemon-reload
    sudo systemctl enable fmp4streamer.service
    sudo systemctl start fmp4streamer.service
    ```

    Watch the logs
    ```
    dmesg
    systemctl status fmp4streamer
    journalctl --system-unit fmp4streamer
    ```



# Viewing
When fmp4streamer.py is running the stream can be viewed from any browser via the following url. **_ip_address_** is the ip address or hostname of your computer, and **_port_** (default: 8000) is the port you set in the configuration section.
```
http://<ip_address>:<port>/
```
  
If you want to view the stream via your favourite media player, you can use the
```
http://<ip_address>:<port>/stream.mp4
```
url.

# Configuration

You can start with the fmp4streamer.conf.dist:

```
cp fmp4streamer.conf.dist fmp4streamer.conf
```

Which contains:
```ini
[server]
listen = 
port = 8000

[/dev/video0]
width = 800
height = 600
fps = 30
# you can set any V4L2 control too, list them with the -l option
h264_profile = High
h264_level = 4.2
h264_i_frame_period = 15
```

You can set all the V4L2 controls via the configuration file. List them with
```
python3 fmp4streamer.py -l
```

# Roadmap
- [x] Use V4L2 instead of raspivid
- [ ] Adding AAC audio to the stream from ALSA
- [ ] Support multiple cameras
- [ ] Rewrite to c or Rust


# Motivation
I wanted to stream my raspberry camera to the browser. I found some solution, but I think they are not optimal or too heavy.

## MJPEG streamers
We have h264, why do we want to stream mjpeg?

## VLC, Gstreamer, FFmpeg
```
apt install vlc
...
0 upgraded, 294 newly installed, 0 to remove and 0 not upgraded.
Need to get 107 MB of archives.
After this operation, 900 MB of additional disk space will be used.
```

```
apt install libgstreamer-plugins-base1.0 gstreamer1.0-tools
...
0 upgraded, 119 newly installed, 0 to remove and 0 not upgraded.
Need to get 51.5 MB of archives.
After this operation, 633 MB of additional disk space will be used.
```

```
apt install ffmpeg
...
0 upgraded, 149 newly installed, 0 to remove and 0 not upgraded.
Need to get 69.3 MB of archives.
After this operation, 714 MB of additional disk space will be used.
```


These are huge. It's bad because lot of binary contains a lot of security bugs and you have to update these packages too in your project lifetime, not just install them in one time.

## [U4VL](https://www.linux-projects.org/uv4l/)
You have to add a custom apt repository to your pi to install, you have to trust these guys. It has closed source components too. So no way.

## [131/h264-live-player](https://github.com/131/h264-live-player)
It uses a baseline javascript h264 decoder. We have better h264 decoders in our hardwares, so why do we want to use a javascript decoder?

## [dans98/pi-h264-to-browser](https://github.com/dans98/pi-h264-to-browser)
I like it, it's good, but I don't like its dependencies
```
apt install python3-picamera
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  libblas3 libgfortran5 liblapack3 python3-numpy
Suggested packages:
  gfortran python-numpy-doc python3-dev python3-pytest python3-numpy-dbg python-picamera-docs
The following NEW packages will be installed:
  libblas3 libgfortran5 liblapack3 python3-numpy python3-picamera
0 upgraded, 5 newly installed, 0 to remove and 0 not upgraded.
Need to get 3,830 kB of archives.
After this operation, 18.1 MB of additional disk space will be used.
```
Do we really need fortran to use a camera in python? Do we look like dinosaurs? NO! And 18.1 MB for some init and some read?

![Fortran dinousaurs](https://i.pinimg.com/564x/7c/3a/c8/7c3ac86430655a9f22b2b2496bb4d1ee.jpg)

```
apt install python3-tornado
...
0 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
Need to get 354 kB of archives.
After this operation, 1,800 kB of additional disk space will be used.
```
It's for the websocket server, not too big but do we really need websockets? No, simple http is enough for streaming to browsers.

[samirkumardas/jmuxer](https://github.com/samirkumardas/jmuxer)

Why do we want to add mp4 headers to h264 with js in the browser? If we want to do it properly we have to send timing information next to h264 too, but what should be the format, custom encoding or protobuf? Or wait, there is already a format for that, I remember it, the mp4! So it should be done on the server side.

[Media Source Extensions](https://caniuse.com/mediasource)

It could be good, but safari on iOS doesn't support it... and why do we need javascript code to play a video?

## [thinkski/berrymse](https://github.com/thinkski/berrymse)

It inspired me, I like it very much, it's simple and minimal, but it uses websockets and the Media Source Extensions...

## fmp4streamer (formerly RaspiWebCam)

So here we are.

It doesn't have any dependencies other than python and V4L2, but they are installed by default.

It adds fMP4 (fragmented MP4) headers to the h264 stream. It uses the python http server to serve it to the browser.

The browsers play it with only one html5 video tag (the most of). No javascript needed.

Safari on iOS plays it only with HLS playlists, but it works too. And the playlists added to the index.html of course.


CHANGELOG.md
# Changelog

## [3.0.4] - 2021-10-31
### Added
- HTTPS/SSL streaming
- Instructions for deployment as a system service (required to access the private key).
- Made fmp4streamer.py "know" that it runs python3 and made it executable
### Fixed
- Spelling error in warning "using deeault config"

## [3.0.3] - 2021-10-25
### Fixed
- Eliminate the rounding errors in sampleduration calculation

## [3.0.2] - 2021-10-24
### Fixed
- Better sampleduration calculation
### Added
- Requesting keyframe at new connections for faster startup

## [3.0.1] - 2021-10-23
### Fixed
- Fixed sampleduration calculation
- Default config shouldn't set h264 profile and level

## [3.0.0] - 2021-10-22
### Added
- Use V4L2 instead of the raspivid, because reading from pipe is slow with big frames
- Config file with ability to set v4l2 parameters
### Changed
- RaspiWebcam has been renamed fmp4streamer, because with V4L2 it works on other linux machines too 

## [2.2.1] - 2021-10-13
### Changed
- Better duration and decode time calculation
- Code, speed and compatibility improvements

## [2.2.0] - 2021-10-10
### Added
- HLS streaming for iPhone
### Changed
- Speed and compatibility improvements

## [2.1.4] - 2021-10-09
### Fixed
- Fixed streaming in Chromium and Android

## [2.1.3] - 2021-10-09
### Changed
- Inline is not needed anymore in the raspivid's parameters
- Speed improvements

## [2.1.2] - 2021-10-09
### Changed
- Works with high fps settings too: use actual decode times for frames, not the fps calculated
### Fixed
- Fix race condition in MP4Writer

## [2.1.1] - 2021-10-09
### Changed
- Speed up frame writing by 4x

## [2.1.0] - 2021-10-08
### Added
- Use a tiny python muxer instead of the javascript based jmuxer.
- Use html5 video tag instead of Fetch api
### Changed
- Use correct timescale

## [2.0.0] - 2021-10-06

- Starting a new project based on [dans98/pi-h264-to-browser](https://github.com/dans98/pi-h264-to-browser).
- Raspiwebcam is the new project name.

### Added
- Use raspivid instead of python3-picamera
- Use Fetch api instead of websockets
### Changed
- Use jmuxer 2.0.2
### Removed
- Removed unnecessary screens
- Removed python3-picamera dependency
- Removed python3-tornado dependency



fmp4streamer.py
#! /usr/bin/python3
import io, socketserver, logging, configparser, getopt, sys, ssl
from threading import Thread
from http import server
from time import time

import bmff
from v4l2camera import V4L2Camera, H264NALU

def get_index_html(codec):
    return f'''<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>fmp4streamer</title>
<link rel="icon" href="data:;base64,iVBORw0KGgo=">
<style>
body {{ margin:0; padding:0; background-color:#303030; }}
#stream {{
  max-height: 100%; max-width: 100%; margin: auto; position: absolute;
  top: 0; left: 0; bottom: 0; right: 0;
}}
</style>
</head>
<body>
  <video controls autoplay muted playsinline id="stream">
    <source src="stream.mp4?{int(time())}" type='video/mp4; codecs="{codec}"'>
    <source src="stream.m3u8" type="application/x-mpegURL">
  </video>
</body>
</html>
'''.encode('utf-8')

def get_stream_m3u8(width, height, codec):
    return f'''#EXTM3U
#EXT-X-STREAM-INF:BANDWIDTH=150000,RESOLUTION={width}x{height},CODECS="{codec}"
streaminf.m3u8
'''.encode('utf-8')

def get_streaminf_m3u8():
    return f'''#EXTM3U
#EXT-X-TARGETDURATION:49057
#EXT-X-VERSION:4
#EXTINF:49057.00,
stream.mp4?{int(time())}
#EXT-X-ENDLIST
'''.encode('utf-8')


class MP4Writer:
    def __init__(self, w, width, height, timescale, sps, pps):
        if not sps or not pps:
            raise ValueError('MP4Writer: sps, pps NALUs are missing!')

        self.seq = 0
        self.sps = sps
        self.pps = pps
        self.start_secs = 0
        self.start_usecs = 0

        self.framebuf = io.BytesIO()

        self.w = w
        self.width = width
        self.height = height
        self.timescale = timescale
        self.timescaleusec = timescale / 1000000
        self.decodetime = 0
        self.prevtime = 0

        self.write_header()


    def write_header(self):
        buf = io.BytesIO()
        bmff.write_ftyp(buf)
        bmff.write_moov(buf, self.width, self.height, self.timescale, self.sps, self.pps)
        self.w.write(buf.getbuffer())

    def add_nalu(self, frame_data, frame_secs, frame_usecs):
        nalutype = H264NALU.get_type(frame_data)

        # our first frame should have SPS, PPS, IDR NALU, so wait until we have an IDR
        if self.seq == 0 and nalutype != H264NALU.IDRTYPE:
            return

        if nalutype == H264NALU.IDRTYPE:
            self.write_frame([self.sps, self.pps, frame_data], frame_secs, frame_usecs, is_idr=True)
        elif nalutype == H264NALU.NONIDRTYPE:
            self.write_frame([frame_data], frame_secs, frame_usecs, is_idr=False)

    def write_frame(self, nalus, frame_secs, frame_usecs, is_idr):
        if self.seq == 0:
            self.start_secs = frame_secs
            self.start_usecs = frame_usecs
            sampleduration = 1
        else:
            sampleduration = round(
                (frame_secs - self.start_secs) * self.timescale +
                (frame_usecs - self.start_usecs) * self.timescaleusec -
                self.decodetime)

        mdatsize = bmff.get_mdat_size(nalus)
        self.framebuf.seek(0)
        bmff.write_moof(self.framebuf, self.seq, mdatsize, is_idr, sampleduration, self.decodetime)
        bmff.write_mdat(self.framebuf, nalus)
        self.w.write(self.framebuf.getbuffer()[:bmff.MOOFSIZE + mdatsize])

        self.seq += 1
        self.decodetime += sampleduration


class StreamingHandler(server.BaseHTTPRequestHandler):
    def do_GET(self):
        if self.path == '/':
            self.send_response(301)
            self.send_header('Location', '/index.html')
            self.end_headers()
        elif self.path == '/index.html':
            self.send_response(200)
            self.send_header('Age', '0')
            self.send_header('Cache-Control', 'no-cache, no-store, must-revalidate')
            self.send_header('Content-Type', 'text/html')
            indexhtml = get_index_html(config.codec())
            self.send_header('Content-Length', len(indexhtml))
            self.end_headers()
            self.wfile.write(indexhtml)
        elif self.path == '/stream.m3u8':
            self.send_response(200)
            self.send_header('Age', '0')
            self.send_header('Cache-Control', 'no-cache, no-store, must-revalidate')
            self.send_header('Content-Type', 'application/x-mpegURL')
            streamm3u8 = get_stream_m3u8(config.width(), config.height(), config.codec())
            self.send_header('Content-Length', len(streamm3u8))
            self.end_headers()
            self.wfile.write(streamm3u8)
        elif self.path == '/streaminf.m3u8':
            self.send_response(200)
            self.send_header('Age', '0')
            self.send_header('Cache-Control', 'no-cache, no-store, must-revalidate')
            self.send_header('Content-Type', 'application/x-mpegURL')
            streaminfm3u8 = get_streaminf_m3u8()
            self.send_header('Content-Length', len(streaminfm3u8))
            self.end_headers()
            self.wfile.write(streaminfm3u8)
        elif self.path.startswith('/stream.mp4'):
            self.send_response(200)
            self.send_header('Age', '0')
            self.send_header('Cache-Control', 'no-cache, no-store, must-revalidate')
            self.send_header('Content-Type', f'video/mp4; codecs="{config.codec()}"')
            self.end_headers()
            try:
                mp4_writer = MP4Writer(self.wfile, config.width(), config.height(), config.timescale(), camera.sps, camera.pps)
                camera.request_key_frame()
                while True:
                    with camera.condition:
                        camera.condition.wait()
                        frame_data = camera.frame_data
                        frame_secs = camera.frame_secs
                        frame_usecs = camera.frame_usecs
                    mp4_writer.add_nalu(frame_data, frame_secs, frame_usecs)
            except Exception as e:
                self.log_message(f'Removed streaming client {self.client_address} {str(e)}')
        else:
            self.send_error(404)
            self.end_headers()


class StreamingServer(socketserver.ThreadingMixIn, server.HTTPServer):
    allow_reuse_address = True
    daemon_threads = True
    def start(self):
        try:
            self.serve_forever()
        except KeyboardInterrupt:
            pass
        finally:
            self.server_close()


class V4L2CameraThread(Thread):
    def __init__(self, camera):
        super(V4L2CameraThread, self).__init__()
        self.camera = camera

    def run(self):
        self.camera.start_capturing()

    def stop(self):
        self.camera.stop_capturing()

class Config(configparser.ConfigParser):
    def __init__(self, configfile):
        super(Config, self).__init__({
            'width': 800,
            'height': 600,
            'fps': 30,
        })
        self.read_dict({'server': {'listen': '', 'port': 8000}})

        if len(self.read(configfile)) == 0:
            logging.warning(f'Couldn\'t read {configfile}, using default config')

        if len(self.sections()) == 1:
            self.add_section('/dev/video0')

        self.device = self.get_devices()[0]

    def get_devices(self):
        return [s for s in self.sections() if s != 'server']

    def get_device(self):
        return self.device

    def width(self):
        return int(self[self.device]['width'])

    def height(self):
        return int(self[self.device]['height'])

    def fps(self):
        return int(self[self.device]['fps'])

    def sampleduration(self):
        return 500

    def timescale(self):
        return self.fps() * self.sampleduration()

    def h264_profile(self):
        return self[self.device].get('h264_profile', 'High')

    def h264_level(self):
        return self[self.device].get('h264_level', '4')

    def codec(self):
        profiles = {'High' : '6400', 'Main': '4d00', 'Baseline': '4200'}
        levels = {'4': '28', '4.1': '29', '4.2': '2a'}
        codec = 'avc1.' + profiles.get(self.h264_profile(), '6400') + levels.get(self.h264_level(), '28')
        return codec

    def keyfile(self):
        return self[self.device].get('keyfile','');

    def certfile(self):
        return self[self.device].get('certfile','');

def usage():
    print(f'usage: python3 {sys.argv[0]} [--help] [--list-controls] [--config CONFIG]\n')
    print(f'optional arguments:')
    print(f'  -h, --help                  show this help message and exit')
    print(f'  -l, --list-controls         list the v4l2 controls and values for the camera')
    print(f'  -c CONFIG, --config CONFIG  use CONFIG as a config file, default: fmp4streamer.conf')


try:
    arguments, values = getopt.getopt(sys.argv[1:], "hlc:", ["help", "list-controls","config="])
except getopt.error as err:
    print(err)
    usage()
    sys.exit(2)

list_controls = False
configfile = "fmp4streamer.conf"

for current_argument, current_value in arguments:
    if current_argument in ('-h', '--help'):
        usage()
        sys.exit(0)
    elif current_argument in ("-l", "--list-controls"):
        list_controls = True
    elif current_argument in ("-c", "--config"):
        configfile = current_value

config = Config(configfile)
device = config.get_device()

camera = V4L2Camera(device, dict(config.items(device)))
if list_controls:
    camera.print_ctrls()
    sys.exit(0)
cameraThread = V4L2CameraThread(camera)
cameraThread.start()

server = StreamingServer((config.get('server', 'listen'), config.getint('server', 'port')), StreamingHandler)
if len(config.get('server', 'keyfile')) != 0 and len(config.get('server', 'certfile')) != 0 :
    server.socket = ssl.wrap_socket(server.socket, \
        config.get('server', 'keyfile'), config.get('server', 'certfile'), server_side=True)
server.start()
cameraThread.stop()
cameraThread.join()
